drivers/tee/optee/optee.c

```diff
@@ -874,15 +967,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
 	return rc;
 }
 
-static int optee_suppl_recv(const struct device *dev, uint32_t func, unsigned int num_params,
+static int optee_suppl_recv(const struct device *dev, uint32_t *func, unsigned int *num_params,
 			    struct tee_param *param)
 {
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req = NULL;
+
+	while (true) {
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		req = (struct optee_supp_req *)sys_dlist_peek_head(&supp->reqs);
+
+		if (req && supp->current == NULL) {
+			if (*num_params < req->num_params) {
+				LOG_ERR("Not enough space for params, need at least %lu",
```

Comment:
it is up to you, because I don't know what will be better
but I think in this case caller should alloc more params and call this function again or you can drop entry from the list and notify the waiting thread

Reply:
With this new logic, the request is left in the queue in case of error, awaiting correct retrieval by the supplicant.

Reply:
ok, got it, I hope it doesn't cause the stall